test/include/seqan3/test/expect_range_eq.hpp

```diff
@@ -0,0 +1,54 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides test utilities for std::ranges::range types.
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <seqan3/std/algorithm>
+#include <seqan3/std/ranges>
+
+#include <seqan3/core/detail/debug_stream_range.hpp>
+#include <seqan3/core/detail/to_string.hpp>
+#include <seqan3/test/pretty_printing.hpp>
+
+namespace seqan3::test
+{
+#define EXPECT_RANGE_EQ(val1, val2) \
+    EXPECT_PRED_FORMAT2(::seqan3::test::expect_range_eq{}, val1, val2);
+
+struct expect_range_eq
+{
+    template <typename rng_t>
+    auto copy_range(rng_t && rng)
+    {
+        using value_t = std::ranges::range_value_t<rng_t>;
+        std::vector<value_t> rng_copy{};
+        std::ranges::copy(rng, std::ranges::back_inserter(rng_copy));
+        return rng_copy;
+    }
+
+    template <std::ranges::range lhs_t, std::ranges::range rhs_t>
```

Comment:
You wrote it in the PR message, but I don't think it would be much work to add another `operator()` that is constrained to `input` or `forward_ranges`?

The standard actually says that `ranges::equal` needs `input_range`s, so why would you require `forward_range`?

Reply:
> You wrote it in the PR message, but I don't think it would be much work to add another `operator()` that is constrained to `input` or `forward_ranges`?

yeah it wouldn't

> The standard actually says that `ranges::equal` needs `input_range`s, so why would you require `forward_range`?

To build the message. I can call an operation on an `input_range` only once.

1) ` if (std::ranges::equal(lhs_copy, rhs_copy))`
2) `::testing::internal::CmpHelperEQFailure(lhs_expression, rhs_expression, lhs_copy, rhs_copy);`

Reply:
I can omit the temporary copy if I can reiterate the same range.

Reply:
Hah right, you need to iterate twice.

Can you omit the copy for ranges where you can do this?

Reply:
I could; Wasn't sure if we wanted this :)

Reply:
Agreed to do later